File: StrStrKMP.java
/**
 * Time Complexity : O(m+n)  where n is the length of the string 1 and m is the length of string 2
 * Space Complexity : O(1) 
 */

public class StrStrKMP {
    public int strStr(String haystack, String needle) {
        if(needle == "" || needle.length() == 0){return 0;}
        int[] table = new int[needle.length()];                                             
        int start =0;                                                                                             
        int m = haystack.length();
        int i = 1;
        table[start] = 0;                                                               
        while(i < needle.length()){
            if(needle.charAt(start) == needle.charAt(i)){                           
                start++;
                table[i] = start;                                                               
                i++;
            } else 
            if(start > 0 && needle.charAt(start) != needle.charAt(i)){                                  
                start = table[start-1];
            } else 
            if(start == 0 && needle.charAt(start) != needle.charAt(i)){                    
                table[i] =  0;
                i++;
            }
        }
        i=0;
        int j = 0;
        while(i < m){                                                                       
            if(haystack.charAt(i) == needle.charAt(j)){
                i++;                                                                                    
                j++;
            }
            if(j == needle.length()){                                                                   
                return i - needle.length();}
            else
            if(i < m && j > 0 && haystack.charAt(i) != needle.charAt(j)){
                j = table[j-1];                                                                 
            } else
            if(i < m && j == 0 && haystack.charAt(i) != needle.charAt(j)){                          
                i++;
            }
        }
        return -1;                                                                          
}
}

File: strStr_KMP.py
class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        '''
        Time Complexity: O(m+n)
        Space Complexity: O(n) 
        '''
        # KMP ALGORITHM
        if(needle == ""):
            return 0
        if(len(haystack)<len(needle)):
            return -1
        '''
        if(len(haystack)==len(needle)):
            if(haystack==needle):
                return 0
            else:
                -1
        '''
        x = [0 for i in range(0,len(needle))]
        j=0
        i=1
        while(i<len(needle)):
            if(needle[i]==needle[j]):
                j+=1
                x[i] = j
                i+=1
            else:
                if(j>0):
                    j = x[j-1]
                else:
                    x[i] = 0
                    i+=1
        
        j=0
        i=0
        
        while(i<len(haystack)):
            if(haystack[i]==needle[j]):
                i+=1
                j+=1
                if(j==len(needle)):
                    return i-j
            else:
                if(j>0):
                    j = x[j-1]
                else:
                    j=0
                    i+=1
        
        return -1
